### DISCUSSION TITLE
Intercity Integration Q&A

### DISCUSSION THREAD

**Discussion by 92shreyansh** [ONDC-Official/mobility-specification#59] [Q&A]
Please use this thread for Q&A related to Intercity Integration

**Comment by Mohitsingh2212** [comment]
hi @92shreyansh ,                                                                                                                                                                                                           invalid city code  in ONDC:TRV12 ,
authorization_header  =   *****
request_data = {
'context':{
'location':{
'country':{
'code':'IND'
}
},
'domain':'ONDC:TRV12',
'timestamp':'2023-11-08T04:41:16.000Z',
'bap_id':'ondc.apnibus.com',
'transaction_id':'a11c1af1-92dd-4aa7-9aa4-7b93efe3909f',
'message_id':'58946676-e16c-4aa7-aee0-ab48637582e1',
'version':'2.0.0',
'action':'search',
'bap_uri':'https://ondc.apnibus.com/bapl',
'ttl':'PT30S'
},
'message':{
'intent':{
'fulfillment':{
'stops':[
{
'type':'START',
'location':{
'descriptor':{
'code':'std:011'
}
},
'time':{
'label':'DATE_OF_JOURNEY',
'timestamp':'2023-11-08T04:41:16.000Z'
}
},
{
'type':'END',
'location':{
'descriptor':{
'code':'std:0522'
}
}
}
],
'vehicle':{
'category':'BUS'
}
},
'payment':{
'tags':[
{
'descriptor':{
'code':'BUYER_FINDER_FEES'
},
'display':False,
'list':[
{
'descriptor':{
'code':'BUYER_FINDER_FEES_PERCENTAGE'
},
'value':'1'
}
]
}
]
}
}
}
}

   when we post request to ondc staging gateway uri then ,                                                                                                                                   
                                                                                                                                                                                                              
  response  =  {
    "context": {
        "domain": "ONDC:TRV12",
        "location": {
            "country": {
                "code": "IND"
            }
        },
        "action": "search",
        "version": "2.0.0",
        "bap_id": "ondc.apnibus.com",
        "bap_uri": "https://ondc.apnibus.com/bapl",
        "transaction_id": "a11c1af1-92dd-4aa7-9aa4-7b93efe3909f",
        "message_id": "58946676-e16c-4aa7-aee0-ab48637582e1",
        "timestamp": "2023-11-08T11:29:34.726Z",
        "ttl": "PT30S"
    },
    "message": {
        "ack": {
            "status": "NACK"
        }
    },
    "error": {
        "type": "Gateway",
        "code": "10002",
        "message": "Throwable: Invalid City Code"
    }
}

**Reply by 92shreyansh** [reply]
Can you please add context.location.city.code = "std:011".

**Reply by Mohitsingh2212** [reply]
hi @92shreyansh 
thanks for the reply,

i getting the error:   
"error": {
        "type": "Gateway",
        "code": "10001",
        "message": "NullPointerException: "
    } 


here is the request body :

{
    "context": {
        "location": {
            "city": {
                "code": "std:011"
            }
        },
        "domain": "ONDC:TRV12",
        "timestamp": "2023-11-17T04:41:16.000Z",
        "bap_id": "ondc.apnibus.com",
        "transaction_id": "9e5d4d19-cfc1-4521-b461-22b21b5982e9",
        "message_id": "d760d3e9-8576-413f-9008-de3d2c7ddfd9",
        "version": "2.0.0",
        "action": "search",
        "bap_uri": "https://ondc.apnibus.com/bapl",
        "ttl": "PT30S"
    },
    "message": {
        "intent": {
            "fulfillment": {
                "stops": [
                    {
                        "type": "START",
                        "location": {
                            "descriptor": {
                                "code": "011"
                            }
                        },
                        "time": {
                            "label": "DATE_OF_JOURNEY",
                            "timestamp": "2023-11-17T04:41:16.000Z"
                        }
                    },
                    {
                        "type": "END",
                        "location": {
                            "descriptor": {
                                "code": "0522"
                            }
                        }
                    }
                ],
                "vehicle": {
                    "category": "BUS"
                }
            },
            "payment": {
                "tags": [
                    {
                        "descriptor": {
                            "code": "BUYER_FINDER_FEES"
                        },
                        "display": false,
                        "list": [
                            {
                                "descriptor": {
                                    "code": "BUYER_FINDER_FEES_PERCENTAGE"
                                },
                                "value": "1"
                            }
                        ]
                    }
                ]
            }
        }
    }
}

and my autherization header is:

"Signature keyId=\"ondc.apnibus.com|uuid****|ed25519\",algorithm=\"ed25519\",created=\"1700198733\",expires=\"1700202333\",headers=\"(created) (expires) digest\",signature=\"****************************""

response body :

{
    "context": {
        "domain": "ONDC:TRV12",
        "location": {
            "city": {
                "code": "std:011"
            }
        },
        "action": "search",
        "version": "2.0.0",
        "bap_id": "ondc.apnibus.com",
        "bap_uri": "https://ondc.apnibus.com/bapl",
        "transaction_id": "7e8fc78a-a540-4e16-a577-2a5ec91a8a4e",
        "message_id": "12c6eca4-eb2a-4893-ab03-4243ef9c8342",
        "timestamp": "2023-11-17T10:22:33.199Z",
        "ttl": "PT30S"
    },
    "message": {
        "ack": {
            "status": "NACK"
        }
    },
    "error": {
        "type": "Gateway",
        "code": "10001",
        "message": "NullPointerException: "
    }
}

**Comment by yashb042** [comment]
@92shreyansh 

While going through the on_search API for intercity, I was confused with some fields -
For example : 

1. Base time of the bus start journey ( times are available for stop wise at fullfillment - START, STOP but there should be a central time which the buyer app shows, definitely should be a part of the response)

2. message.catalog.providers.items.price.currency 
    message.catalog.providers.items.price.value
Is this supposed to be an array of prices for different types of seats? Following is the screenshot for sample response provided 
<img width="547" alt="Screenshot 2023-11-09 at 3 30 33 PM" src="https://github.com/ONDC-Official/mobility-specification/assets/67835516/beeec997-a047-44a7-b5ab-d8651380da31">


For your reference, I am attaching a response we used here at in-house ONDC for redBus 
It contains clear cut segregation between boarding and dropping points, a standard bus departure/arrival time, different types of prices and their clear breakups. 

[https://docs.google.com/document/d/1w_nP50nO8wAjre0UXCxpkZdcczEqQQe2s-d67uUoKss/edit](url)

If possible, please improve the responses for the above 2 fields and compare the rest of fields with redBus search response which we give to other buyers.

**Comment by shivapnibus** [comment]
Hi @92shreyansh .

previously I was subscribed to "nic2004:52110" domain by mistake and later the ONDC team has changed my subscription to my desired domain "ONDC:TRV12" upon raising an issue in this forum. 

Post that I could successfully see my self subscribed to the "ONDC:TRV12" domain during the /lookup call in the staging registry.

Here are my details after lookup call:

{
        "subscriber_id": "ondc.apnibus.com",
        "status": "SUBSCRIBED",
        "ukId": " 9bdac6c4-56b4-4455-98f9-6a8eb14177f7",
        "subscriber_url": "https://ondc.apnibus.com/bapl",
        "country": "IND",
        "domain": "ONDC:TRV12",
        "valid_from": "2023-10-23T10:44:54.101Z",
        "valid_until": "2023-10-25T13:44:54.101Z",
        "type": "BAP",
        "signing_public_key": "oBS6nMNeTMQdGT352EpwjgteNyXCDy9FR+WSEKfmwAQ=",
        "encr_public_key": "MCowBQYDK2VuAyEABq3k6L/ZY5NQx5pxOPU5TYGPkxAwPFU7jb9OEcVIzHo=",
        "created": "2023-10-23T09:00:06.241Z",
        "updated": "2023-10-23T09:00:06.241Z",
        "br_id": " 9bdac6c4-56b4-4455-98f9-6a8eb14177f7",
        "city": "std:011"
    },
    
    
   **NOTE**: I just received the /on_subscribe callback once during the successful registration to "nic2004:52110" domain. Later when the ONDC team had changed my domain to "ONDC:TRV12" from their backend, i didn't receive the /on_subscribe callback for this change.
 As a result, my signing Key Pairs have not changed since my first subscription.
   
   **Now, when i hit the /search endpoint, I'm getting the below error. Not Sure if the above case is the reason for this:**
   
**request body to /search endpoint:**
{
    "context": {
        "location": {
            "city": {
                "code": "std:011"
            }
        },
        "domain": "ONDC:TRV12",
        "timestamp": "2023-11-18T17:51:10.771Z",
        "bap_id": "ondc.apnibus.com",
        "transaction_id": "9e5d4d19-cfc1-4521-b461-22b21b5982e9",
        "message_id": "d760d3e9-8576-413f-9008-de3d2c7ddfd9",
        "version": "2.0.0",
        "action": "search",
        "bap_uri": "https://ondc.apnibus.com/bapl",
        "ttl": "PT30S"
    },
    "message": {
        "intent": {
            "fulfillment": {
                "stops": [
                    {
                        "type": "START",
                        "location": {
                            "descriptor": {
                                "code": "011"
                            }
                        },
                        "time": {
                            "label": "DATE_OF_JOURNEY",
                            "timestamp":"2023-11-18T18:51:10.771Z"
                        }
                    },
                    {
                        "type": "END",
                        "location": {
                            "descriptor": {
                                "code": "0522"
                            }
                        }
                    }
                ],
                "vehicle": {
                    "category": "BUS"
                }
            },
            "payment": {
                "tags": [
                    {
                        "descriptor": {
                            "code": "BUYER_FINDER_FEES"
                        },
                        "display": false,
                        "list": [
                            {
                                "descriptor": {
                                    "code": "BUYER_FINDER_FEES_PERCENTAGE"
                                },
                                "value": "1"
                            }
                        ]
                    }
                ]
            }
        }
    }
}

**Headers:** 
Signed using the private_key. I have followed all the steps in the header creation doc, and passed the below header that contains the signed request body text signature.

Signature keyId="ondc.apnibus.com|9bdac6c4-56b4-4455-98f9-6a8eb14177f7|ed25519",algorithm="ed25519",created="1700202101",expires="1700205701",headers="(created) (expires) digest",signature="8D/S/uRoX5xZM4o0kXMJK9TaHiavOrLeKbq6/sgMeVldOsQ+DLp/8LklCMJFYIt4cJNUKHGtd0it3jn2jwq8Dg=="


**Response:**
{
    "context": {
        "domain": "ONDC:TRV12",
        "location": {
            "city": {
                "code": "std:011"
            }
        },
        "action": "search",
        "version": "2.0.0",
        "bap_id": "ondc.apnibus.com",
        "bap_uri": "https://ondc.apnibus.com/bapl",
        "transaction_id": "9e5d4d19-cfc1-4521-b461-22b21b5982e9",
        "message_id": "d760d3e9-8576-413f-9008-de3d2c7ddfd9",
        "timestamp": "2023-11-18T12:45:04.549Z",
        "ttl": "PT30S"
    },
    "message": {
        "ack": {
            "status": "NACK"
        }
    },
    "error": {
        "type": "Gateway",
        "code": "10001",
        "message": "NullPointerException"
    }
}

----------------------

We are unable to carry further implementations and tests and stuck here.

Please help us resolve this so that we can proceed further.

**Reply by shivapnibus** [reply]
Would be great if you could help us in this @92shreyansh

**Comment by praveen-intrcity** [comment]
Hi, 
Can anyone from the buyer's end share the request payload and Authorization header for the search request.
So that we test the search <=> on_search flow for search 1 and search 2.

**Comment by praveen-intrcity** [comment]
We are done with all search apis (1, 2 and 3)on stage.

**Reply by Mohitsingh2212** [reply]
@praveen-intrcity what do you mean by search apis 1, 2, and 3. We just have /search and /on_search right?

**Reply by praveen-intrcity** [reply]
There are 3 types of search/on_search apis as described in the swagger (Flow UI) .
![Screenshot from 2023-12-06 13-16-12](https://github.com/ONDC-Official/mobility-specification/assets/151712034/6062fe0c-a9c2-4151-b371-85ce5b12dc67)

**Reply by Mohitsingh2212** [reply]
@praveen-intrcity  
 I noticed I'm not receiving callbacks. Could you confirm if your server is currently live. when I hit my search API, I get an ack, but I'm not receiving any callbacks if your server is live, could you provide me with its current status

**Reply by praveen-intrcity** [reply]
Server is up now. You can try now.

**Reply by Mohitsingh2212** [reply]
@praveen-intrcity 
Thanks for confirming that the server is up. I'm still not receiving callbacks. Could you please check if everything is configured correctly for callbacks on your end? Let me know if there's anything specific I should be looking into

**Reply by praveen-intrcity** [reply]
@Mohitsingh2212 
Please Check the DATE_OF_JOURNEY 

{"context":{"location":{"city":{"code":"std:011"}},"domain":"ONDC:TRV12","timestamp":"2023-12-08T08:36:16.9254","bap_id":"ondc.apnibus.com","transaction_id":"08c28c7f-390d-46d3-9695-80cd04ad6f3c","message_id":"0b61f058-238d-41d9-b03e-6b43a8018579","version":"2.0.0","action":"search","bap_uri":"https://ondc.apnibus.com/bapl","ttl":"PT3S"},"message":{"intent":{"fulfillment":{"stops":[{"type":"START","location":{"descriptor":{"code":"std:011"}},"time":{"label":"DATE_OF_JOURNEY","timestamp":"2023-12-0T00:00:00.000Z"}},{"type":"END","location":{"descriptor":{"code":"std:0522"}}}],"vehicle":{"category":"BUS"}},"payment":{"tags":[{"descriptor":{"code":"BUYER_FINDER_FEES"},"display":false,"list":[{"descriptor":{"code":"BUYER_FINDER_FEES_PERCENTAGE"},"value":"1"}]}]}}}}

**Reply by praveen-intrcity** [reply]
@Mohitsingh2212

Since we have enable the signature verification.
if the signature is valid then you will  receive the callback.
otherwise you will receive an error message in the acknowledgement.

**Reply by Mohitsingh2212** [reply]
@praveen-intrcity 
I've reviewed the acknowledgment message, and it appears that the status is "ACK," indicating that the signature verification was successful. However, I haven't received any callback yet. To ensure clarity, could you please confirm if there are specific details or parameters I should be mindful of in the callback payload?

Additionally, if there's anything specific you'd like me to check or if you have any suggestions on how to proceed, please let me know.

**Reply by praveen-intrcity** [reply]
The response message  return the status "NACK" .
{"message":{{ack: {"status":"NACK","tags":[]}},"error":{"code":"","paths":"","message":"Invalid Signature"}}

Currently we have disabled the signature verification. so you that you can receive the callback and do the further.
we will enable this later.

For signature part. you can try to verify the generated signature at your end also.

**Reply by Mohitsingh2212** [reply]
@praveen-intrcity 
I appreciate your update and the clarification regarding signature verification. After conducting tests on our APIs without header verification, we have observed smooth functionality on our end.

To reiterate, we post requests to the ONDC gateway, receive acknowledgment messages, and successfully process the requests. However, it seems that the ONDC gateway may not be verifying headers before broadcasting the requests to us. Consequently, we receive the post request without header verification.

In light of this, it would be beneficial for us to engage with the ONDC team to discuss and clarify the situation. It is crucial to ensure seamless alignment between our systems and determine whether ONDC expects us to perform header verification on our end or if they will handle it at their level.

Could you please take the initiative to coordinate with the ONDC team to obtain clarification on this matter? Once we have a clear understanding, we can proceed accordingly and make any necessary adjustments

**Comment by Mukul-Mathur** [comment]
hi @92shreyansh 

Could you suggest link for document cancel API with enum value. I am referring to this link https://ondc-official.github.io/mobility-specification/#/Provider%20Platform/post_cancel . I see below input params but no enum value. 
<img width="862" alt="Screenshot 2023-12-07 at 3 32 25 PM" src="https://github.com/ONDC-Official/mobility-specification/assets/153170297/f15474d5-9173-4ee8-8e47-eb568478502e">

Also had queries on how to access staging environment. I understand we have to fill in request [here](https://docs.google.com/forms/d/1k5k-N2JW4azLsdkJVbWjlsW549Nz5tUatYozSmJERQk/viewform?edit_requested=true) , but post that how do we get links to seller staging APIs ? 

Regards
Mukul
